File: bleDataReceived.m
% Define a callback function to handle incoming data
function bleDataReceived(src, ~)
    
    global finished
    global first
    global elapsedTime
    global characteristicData
    global count
    global count_last
    global numBytesReceived
    global c_write
    global c_read
    BUFFER_LENGTH=244; 
    datacounter=0;
    datagenerated=[];
    last_value1=0;
    first_send=1;
    if (first==1)
        tic;
        first=0;
    end
    if finished==0
        newData = read(src, 'latest');
        last_value1=newData(end);
        last_value2=newData(end-1);
        numBytesReceived = numBytesReceived + size(newData,2);
        if size(newData,2)<244
           paddingSize = 244 - size(newData,2);
           zerorow=zeros(1,paddingSize);
           newData = [newData, zerorow]; 
        end
        characteristicData = [characteristicData; newData];
    end
    
    % Check for the end of transmission
    if numBytesReceived >= 2 && last_value1 == 255 && last_value2 == 255
        finished=1;
        disp('Data collection finished.');
        unsubscribe(c_read);
        elapsedTime = toc;
        % compute throughput
        numBitsReceived = numBytesReceived * 8; % Assuming each value is 1 byte (8 bits)
        throughput_receive = numBitsReceived / (elapsedTime*1000) % Bits per milliseconds
        NUMBER_OF_CYCLES = floor(numBytesReceived/ BUFFER_LENGTH);
        REMAINING_8BITS_PACKETS = mod(numBytesReceived, BUFFER_LENGTH);
        data_received=zeros(1, (NUMBER_OF_CYCLES+1)*194+153);
        for i = 1:NUMBER_OF_CYCLES
            for j = 1:5:236
                data_received(count + 1) = bitor(bitshift(uint16(characteristicData(i,j)), 2), bitshift(uint16(characteristicData(i,j+1)), -6));
                data_received(count + 2) = bitor(bitshift(bitand(uint16(characteristicData(i,j+1)), uint16(0x3F)), 4), bitshift(bitand(uint16(characteristicData(i,j+2)), uint16(0xF0)), -4));
                data_received(count + 3) = bitor(bitshift(bitand(uint16(characteristicData(i,j+2)), uint16(0x0F)), 6), bitshift(bitand(uint16(characteristicData(i,j+3)), uint16(0xFC)), -2));
                data_received(count + 4) = bitor(bitshift(bitand(uint16(characteristicData(i,j+3)), uint16(0x03)), 8), uint16(characteristicData(i,j+4)));
                count = count + 4;
            end
            for j = 241:2:BUFFER_LENGTH
                count = count + 1;
                data_received(count) = bitor(bitshift(uint16(characteristicData(i,j+1)), 8), uint16(characteristicData(i,j)));
            end
        end
        i = NUMBER_OF_CYCLES+1;
        for j = 1:5:(REMAINING_8BITS_PACKETS-6)
            data_received(count + 1) = bitor(bitshift(uint16(characteristicData(i,j)), 2), bitshift(uint16(characteristicData(i,j+1)), -6));
            data_received(count + 2) = bitor(bitshift(bitand(uint16(characteristicData(i,j+1)), uint16(0x3F)), 4), bitshift(bitand(uint16(characteristicData(i,j+2)), uint16(0xF0)), -4));
            data_received(count + 3) = bitor(bitshift(bitand(uint16(characteristicData(i,j+2)), uint16(0x0F)), 6), bitshift(bitand(uint16(characteristicData(i,j+3)), uint16(0xFC)), -2));
            data_received(count + 4) = bitor(bitshift(bitand(uint16(characteristicData(i,j+3)), uint16(0x03)), 8), uint16(characteristicData(i,j+4)));
            count = count + 4;
            count_last = j+4;
        end
        count_last = count_last + 1;
        for j = count_last:2:REMAINING_8BITS_PACKETS+1
            count = count + 1;
            data_received(count) = bitor(bitshift(uint16(characteristicData(i,j+1)), 8), uint16(characteristicData(i,j)));
        end
        assignin('base', 'exported_data_received', data_received);
          %% send data
        for i=1:32400
            if datacounter>253
                datacounter=0;
            else 
                datacounter=datacounter+1;
            end
            if i==32400-1 || i==32400
                datacounter=255;
            end
            datagenerated=[datagenerated, datacounter];
            if size(datagenerated, 2)==BUFFER_LENGTH || (i==32400)
                write(c_write, datagenerated,'WithoutResponse');
                if first_send==1
                   first_send=0;
                   tic
                end
                if i==32400
                   elapsedTime_send=toc;
                   throughput_send=i*8/(elapsedTime_send*1000)
                end
                datagenerated=[];
           end
        end
        first=1;
        finished=0;
        numBytesReceived=0;
        subscribe(c_read);
        c_read.DataAvailableFcn = @bleDataReceived;
    end 
end

File: bleDataReceived.py
import time
from pygatt import BLEDevice, Characteristic, exceptions

# settings
serviceUUID_write = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
characteristicUUID_write = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
serviceUUID_read = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
characteristicUUID_read = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"

# # Connect to BLE device
# try:
#     device = BLEDevice("D238532C0635")
#     device.connect()
#     print("Connected to BLE device")
# except exceptions.NotConnectedError:
#     print("Failed to connect to BLE device")

# # Find the characteristic objects for writing and reading data
# c_write = device.char_write(serviceUUID_write, characteristicUUID_write)
# c_read = device.char_read(serviceUUID_read, characteristicUUID_read)


# Define callback function for handling received data
def ble_data_received(
    handle,
    value,
    first,
    finished,
    numBytesReceived,
    characteristicData,
    c_read,
    device,
    c_write,
):
    # global finished, first, elapsedTime, characteristicData, count, count_last, numBytesReceived, c_write, c_read
    BUFFER_LENGTH = 244
    datacounter = 0
    datagenerated = []
    last_value1 = 0
    first_send = 1

    if first == 1:
        start_time = time.time()
        first = 0

    if finished == 0:
        newData = value
        last_value1 = newData[-1]
        last_value2 = newData[-2]
        numBytesReceived += len(newData)

        if len(newData) < 244:
            paddingSize = 244 - len(newData)
            zerorow = [0] * paddingSize
            newData += zerorow

        characteristicData.append(newData)

    # Check for the end of transmission
    if numBytesReceived >= 2 and last_value1 == 255 and last_value2 == 255:
        finished = 1
        print("Data collection finished.")
        elapsedTime = time.time() - start_time
        throughput_receive = (
            numBytesReceived * 8 / (elapsedTime * 1000)
        )  # Bits per milliseconds

        NUMBER_OF_CYCLES = numBytesReceived // BUFFER_LENGTH
        REMAINING_8BITS_PACKETS = numBytesReceived % BUFFER_LENGTH
        data_received = []

        for i in range(NUMBER_OF_CYCLES):
            for j in range(0, 236, 5):
                data_received.append(
                    (characteristicData[i][j] << 2)
                    | (characteristicData[i][j + 1] >> 6)
                )
                data_received.append(
                    ((characteristicData[i][j + 1] & 0x3F) << 4)
                    | (characteristicData[i][j + 2] >> 4)
                )
                data_received.append(
                    ((characteristicData[i][j + 2] & 0x0F) << 6)
                    | (characteristicData[i][j + 3] >> 2)
                )
                data_received.append(
                    ((characteristicData[i][j + 3] & 0x03) << 8)
                    | characteristicData[i][j + 4]
                )

            for j in range(241, BUFFER_LENGTH, 2):
                data_received.append(
                    (characteristicData[i][j + 1] << 8) | characteristicData[i][j]
                )

        for j in range(0, REMAINING_8BITS_PACKETS - 6, 5):
            data_received.append(
                (characteristicData[NUMBER_OF_CYCLES][j] << 2)
                | (characteristicData[NUMBER_OF_CYCLES][j + 1] >> 6)
            )
            data_received.append(
                ((characteristicData[NUMBER_OF_CYCLES][j + 1] & 0x3F) << 4)
                | (characteristicData[NUMBER_OF_CYCLES][j + 2] >> 4)
            )
            data_received.append(
                ((characteristicData[NUMBER_OF_CYCLES][j + 2] & 0x0F) << 6)
                | (characteristicData[NUMBER_OF_CYCLES][j + 3] >> 2)
            )
            data_received.append(
                ((characteristicData[NUMBER_OF_CYCLES][j + 3] & 0x03) << 8)
                | characteristicData[NUMBER_OF_CYCLES][j + 4]
            )

        for j in range(REMAINING_8BITS_PACKETS - 6, REMAINING_8BITS_PACKETS + 1, 2):
            data_received.append(
                (characteristicData[NUMBER_OF_CYCLES][j + 1] << 8)
                | characteristicData[NUMBER_OF_CYCLES][j]
            )

        print("Throughput (receive):", throughput_receive)
        print("Data received:", data_received)

        # device.unsubscribe(characteristicUUID_read)
        characteristicData = []
        count = 0
        count_last = 0
        numBytesReceived = 0
        # subscribe(c_read)
        # device.unsubscribe(characteristicUUID_read)
        # device.disconnect()
        c_read.char_read_callback = ble_data_received
